Bind the serialize handler to the correct "this"

boris-petrov opened the pull request:
I hit this issue in an Ember application where `this` in the `serialize` hook inside a router was undefined. I guess this is a bug.

Not sure whether the same should be done for `this.serializer` on the line that I've changed. I'm also not sure that I've handled the `async` case in the tests correctly - I just ignore it.

I'm willing to fix anything that you don't like in this PR. :)

stefanpenner commented on lib/router/handler-info/unresolved-handler-info-by-object.js:
```diff
@@ -25,7 +25,7 @@ var UnresolvedHandlerInfoByObject = subclass(HandlerInfo, {
     var model = _model || this.context,
         names = this.names,
-        serializer = this.serializer || (this.handler && this.handler.serialize);
+        serializer = this.serializer || (this.handler && this.handler.serialize && this.handler.serialize.bind(this.handler));
```
how does this get unbound?

boris-petrov replied:
That's the way JavaScript works. When you get a reference to a method, `this` in it is not bound. The same applies for `this.serializer`, but I couldn't figure out where you use/set that and that's why I left it unmodified.

stefanpenner replied:
Oh sorry, i totally didn't read the code correctly. You are correct!

stefanpenner replied:
I should not be allowed to comment pre-coffee...

stefanpenner replied:
@trentmwillis it looks like you added this, was this intentional to decoupling serialize from the `this` re: lazy engines? Or an oversight?

stefanpenner replied:
We should add a test regardless, ensuring behavior remains as we want.

stefanpenner commented:
> this.serializer is intentionally unbound though.

We should then likely change this PR to test `this.serialize` is unbound.

boris-petrov commented:
I pushed another commit with another test to check that `this.serializer` is unbound. Please let me know if you need any more changes.

spectras commented:
I just stumbled upon this. For what it's worth, invoking route's `serialize` method with no `this` breaks the use of mixins in Ember routes, as `superWrapper` looks for `this._super`.

It also prevents having a base route object with some generic methods that you can extend and just set a few attributes.

Minimalist example hitting both:
```js
// my-mixin.js
export default Ember.Mixin.create({
    serialize: function (model, params) {
        const key = get(this, 'modelName') + '_id';
        if (get(model, 'isNew')) {
            let result = {};
            result[key] = 'new';
            return result;
        }
        return this._super(model, params);
    }
});
```
```js
// my-route.js
export default Ember.Route.extend(MyMixin, {
    modelName: 'myModel'
});
```

rwjblue commented on lib/router/handler-info/unresolved-handler-info-by-object.js:
```diff
@@ -25,7 +25,8 @@ var UnresolvedHandlerInfoByObject = subclass(HandlerInfo, {
         names = this.names,
-        serializer = this.serializer || (this.handler && this.handler.serialize);
+        // this.serializer is intentionally unbound since getSerializer is expected to return a stateless function
+        serializer = this.serializer || (this.handler && this.handler.serialize && this.handler.serialize.bind(this.handler));
```
I would prefer to refactor this a bit to avoid needing to `.bind` again for every call to the same handler's `serialize`.  I think the simplest way woul dbe to remove this `serializer` local variable, and update the `if` conditional below to:

```js
if (this.serializer) {
  // invoke this.serializer unbound (getSerializer returns a stateless function)
  return (0, this.serializer)(model, names);
} else if (this.handler && this.handler.serialize) {
  return this.handler.serialize(model, names);
}
```

boris-petrov commented:
@rwjblue - I did the changes you requested, you're right that it is much better that way. However there's a JSHint error which I'm not sure how to resolve:

`router/handler-info/unresolved-handler-info-by-object.js: line 38, col 34, Bad invocation.`

This is on the line:

```javascript
return (0, this.serializer)(model, names);
```

boris-petrov commented:
@trentmwillis - thanks for the tip, I used the `.call` version - I prefer it.

@rwjblue - I did the changes as requested, rebased on master and fixed the tests. Please merge this soon as I don't want to rebase again after a million changes. :D

Turbo87 commented:
FWIW the following workaround in the relevant routes seems to work fine if you're stuck on older Ember versions:

```js
  init() {
    this._super(...arguments);

    // see https://github.com/tildeio/router.js/pull/214
    this.serialize = this.serialize.bind(this);
  },
```

rwjblue commented:
Thank you so much @boris-petrov!
